src/vivliostyle/layoututil.js

```diff
@@ -397,5 +399,183 @@ goog.scope(function() {
+        this.headerElement = element;
+    };
+    /**
+     * @param {Element} element
+     */
+    RepetitiveElements.prototype.setFooterElement = function(element) {
+        if (this.footerElement) return; // use first one.
+        this.footerElement = element;
+    };
+
+    /**
+     * @param {!adapt.vtree.ClientLayout} clientLayout
+     */
+    RepetitiveElements.prototype.updateHeight = function(clientLayout) {
+        if (this.headerElement) {
+            this.headerHeight = this.getElementHeight(this.headerElement, clientLayout);
+        }
+        if (this.footerElement) {
+            this.footerHeight = this.getElementHeight(this.footerElement, clientLayout);
+        }
+    };
+
+    /**
+     * @private
+     * @param {!Element} element
+     * @param {!adapt.vtree.ClientLayout} clientLayout
+     */
+    RepetitiveElements.prototype.getElementHeight = function(element, clientLayout) {
+        var rect = clientLayout.getElementClientRect(element);
+        return this.vertical ? rect["width"] : rect["height"];
+    };
+
+    RepetitiveElements.prototype.prepareLayoutFragment = function() {
+        this.previousPenalty  = null;
+        this.isSkipHeader = this.isSkipFooter = false;
+        this.headerViewNodes = [];
+        this.footerViewNodes = [];
+        this.enableStatusUpdate = true;
+        this.enableSkippingFooter = true;
+        this.enableSkippingHeader = true;
+    };
+    /**
+     * @param {!Element} rootViewNode
+     * @param {?Node} firstChild
+     */
+    RepetitiveElements.prototype.appendHeaderToFragment = function(rootViewNode, firstChild) {
+        if (!this.headerElement) return;
+        var headerViewNode = this.headerElement.cloneNode(true);
+        this.headerViewNodes.push(headerViewNode);
+        rootViewNode.insertBefore(headerViewNode, firstChild);
+    };
+    /**
+     * @param {!Element} rootViewNode
+     * @param {?Node} firstChild
+     */
+    RepetitiveElements.prototype.appendFooterToFragment = function(rootViewNode, firstChild) {
+        if (!this.footerElement) return;
+        var footerViewNode = this.footerElement.cloneNode(true);
+        this.footerViewNodes.push(footerViewNode);
+        rootViewNode.insertBefore(footerViewNode, firstChild);
+    };
+
+    /**
+     * @return {number}
+     */
+    RepetitiveElements.prototype.getPenalty = function() {
+        return (this.isSkipHeader ? 1 : 0)
+             + (this.isSkipFooter ? 1 : 0);
+    };
+
+    /**
+     * @return {number}
+     */
+    RepetitiveElements.prototype.getNextPenaltyIncreasement = function() {
+        if (!this.enableStatusUpdate) return 0;
+        if ((!this.isSkipFooter && this.enableSkippingFooter)
+        || (!this.isSkipHeader && this.enableSkippingHeader)) {
+            return 1;
+        } else {
+            return 0;
+        }
+    };
+
+    /**
+     * @return {number}
+     */
+    RepetitiveElements.prototype.calculateElementHeight = function() {
```

Comment:
Is `calculateElementHeight` an appropriate name?
It seems to me that the value calculated here is some kind of offset, not height of elements.